pkg/storage/sync/storage.go

```diff
@@ -0,0 +1,165 @@
+type SyncStorage struct {
+	storage.Storage
+	storages     []storage.Storage
+	eventStream  watch.AssociatedEventStream
+	updateStream UpdateStream
+	monitor      *watch.Monitor
+}
+
+var _ storage.Storage = &SyncStorage{}
+
+// NewSyncStorage constructs a new SyncStorage
+func NewSyncStorage(rwStorage storage.Storage, wStorages ...storage.Storage) storage.Storage {
+	ss := &SyncStorage{
+		Storage:      rwStorage,
+		storages:     append(wStorages, rwStorage),
+		updateStream: make(UpdateStream, updateBuffer),
+	}
+
+	for _, s := range ss.storages {
+		if watchStorage, ok := s.(watch.WatchStorage); ok {
+			watchStorage.SetEventStream(ss.getEventStream())
+		}
+	}
+
+	if ss.eventStream != nil {
+		ss.monitor = watch.RunMonitor(ss.monitorFunc)
+	}
+
+	return ss
+}
+
+func (ss *SyncStorage) getEventStream() watch.AssociatedEventStream {
+	if ss.eventStream == nil {
+		ss.eventStream = make(watch.AssociatedEventStream)
+	}
+
+	return ss.eventStream
+}
+
+// Set is propagated to all Storages
+func (ss *SyncStorage) Set(gvk schema.GroupVersionKind, obj meta.Object) error {
+	return ss.runAll(func(s storage.Storage) error {
+		return s.Set(gvk, obj)
+	})
+}
+
+// Patch is propagated to all Storages
+func (ss *SyncStorage) Patch(gvk schema.GroupVersionKind, uid meta.UID, patch []byte) error {
+	return ss.runAll(func(s storage.Storage) error {
+		return s.Patch(gvk, uid, patch)
+	})
+}
+
+// Delete is propagated to all Storages
+func (ss *SyncStorage) Delete(gvk schema.GroupVersionKind, uid meta.UID) error {
+	return ss.runAll(func(s storage.Storage) error {
+		return s.Delete(gvk, uid)
+	})
+}
+
+func (ss *SyncStorage) Close() {
+	// Close all WatchStorages
+	for _, s := range ss.storages {
+		if watchStorage, ok := s.(watch.WatchStorage); ok {
+			watchStorage.Close()
+		}
+	}
+
+	close(ss.eventStream) // Close the event stream
+	ss.monitor.Wait()     // Wait for the monitor goroutine
+}
+
+func (ss *SyncStorage) GetUpdateStream() UpdateStream {
+	return ss.updateStream
+}
+
+// runAll runs the given function for all Storages in parallel and aggregates all errors
+func (ss *SyncStorage) runAll(f func(storage.Storage) error) (err error) {
+	type result struct {
+		int
+		error
+	}
+
+	errC := make(chan result)
+	for i, s := range ss.storages {
+		go func(i int, s storage.Storage) {
+			errC <- result{i, f(s)}
+		}(i, s) // NOTE: This requires i and s as arguments, otherwise they will be evaluated for one Storage only
+	}
+
+	for i := 0; i < len(ss.storages); i++ {
+		if result := <-errC; result.error != nil {
+			if err == nil {
+				err = fmt.Errorf("SyncStorage: error in Storage %d: %v", result.int, result.error)
+			} else {
+				err = fmt.Errorf("%v\n%29s %d: %v", err, "and error in Storage", result.int, result.error)
+			}
+		}
+	}
+
+	return
+}
+
+func (ss *SyncStorage) monitorFunc() {
+	defer log.Debug("SyncStorage: monitoring thread stopped")
+
+	// This is the internal client for propagating updates
+	c := client.NewClient(ss)
```

Comment:
is this run only once, or multiple times?  if once, then cache the client in the syncstorage object

Reply:
It's run once, in a goroutine. Caching it in SyncStorage brings no benefits, as it's only used in monitorFunc and doesn't get removed.